Merging content about importing from relational into Neo4j (#530)
neo4j/docs-getting-started: modules/ROOT/pages/data-import/relational-to-graph-import.adoc

[[relational-to-graph-import]]
= Import from a relational database into Neo4j
:tags: data-import, graph-import, relational-graph, load-csv, apoc, etl-tool, kettle, driver-import
:description: This page shows the different ways you can import data from a relational database to Neo4j. Completing this guide will give you the tools to choose how to import your relational data and transform it to the graph.

[TIP]
====

> It is recommended that you create and understand your xref:data-modeling/index.adoc[graph data model] before transferring the data from an existing relational structure to a graph.

Reviewer: ```suggestion
You need to have a clear understanding of your xref:data-modeling/index.adoc[graph data model] before transferring the data from an existing relational structure to a graph.
```
This is not just recommended, if you don't have a data model, you can't import, AFAIK.

====

Depending on the environment you are working in, the format of the exported dataset, and the link:https://neo4j.com/docs/deployment-options/[Neo4j deployment method] of your choice, there are different ways you can import data from a relational database into a graph.

This page lists some of the recommended methods.
For a full list, refer to xref:data-import/index.adoc[].

> == Aura Import

Reviewer: ```suggestion
== Import tool
```
Since it's not the only way to import data into an Aura instance. Also, it reads less awkwardly when we mention the standalone tool.

link:https://neo4j.com/docs/aura/import/introduction/[Import] is a tool for importing data into an Aura instance.
Ideal to get started quickly with testing and prototyping, it is also available as the stand-alone tool link:https://neo4j.com/docs/data-importer/current/[Data Importer].

=== Data sources

image::data-sources.svg[Screenshot of Aura Import's feature Data Sources,width=700,role=popup]

Through the link:https://neo4j.com/docs/aura/import/data-sources/[Data sources] tab, you can provide remote access to data stored in the following sources:

* PostgreSQL
* MySQL
* SQL Server
* Oracle
* BigQuery
* Databricks
* Redshift
* Snowflake
* Azure Synapse
* AWS S3
* Azure Blobs & Data Lake Storage
* Google Cloud Storage

> Refer to link:https://neo4j.com/docs/aura/import/file-provision/[Aura -> File provision] for a full guide on the tool.

Reviewer: ```suggestion
Refer to link:https://neo4j.com/docs/aura/import/introduction/[Aura -> Import] for a full guide on the tool.

Additionally, the Import tool also allows you to stream local _CSV_ files.
```

== `LOAD CSV`

A common format that many systems can handle is a flat file of comma-separated values (CSV).
After exporting your relational data into a CSV file, you can load it into a Neo4j instance using the Cypher command `LOAD CSV`.

This built-in Cypher function allows users to take existing or exported CSV files and load them into Neo4j with Cypher statements to read, transform, and import the data to the graph database.
It allows the user to run statements individually or run them batched in a Cypher script.

> See xref:data-import/csv-import.adoc[] for a tutorial on how to use `LOAD CSV` or refer to xref:data-modeling/relational-to-graph-modeling.adoc[] for specific guidance on how to model data from relational to graph.

Reviewer: ```suggestion
See xref:data-import/csv-import.adoc[] for a tutorial on how to use `LOAD CSV`.
Refer to xref:data-modeling/relational-to-graph-modeling.adoc[] for specific guidance on how to model data from relational to graph.
```

> In case you are dealing with bigger data sets, you may prefer to use the `neo4j-admin database import` command instead.

Reviewer: ```suggestion
When you are dealing with bigger data sets (>10 million rows) and if you can afford downtime, you can use the `neo4j-admin database import` command instead.
```
Maybe more information about this option here?

See link:https://neo4j.com/docs/operations-manual/current/import/[Operations -> Import] for more details.

== APOC

link:https://neo4j.com/docs/apoc/current/import/[APOC] is Neo4j's utility library for handling data import, as well as data transformations and manipulations.
From converting values to altering the data model, this library allows you to combine and chain procedures in order to get the desired results.

> You can use APOC to import files or data from a URL in CSV, JSON, XML, and XLS, as well as loading data straight from a database (using JDBC).

Reviewer: Do we use _italics_ for file formats?

APOC also supports loading data from link:https://neo4j.com/docs/apoc/current/import/web-apis/[web APIs] and link:https://neo4j.com/docs/apoc/current/import/import-graphml/[GraphML].

When you call these procedures, you can pass in the data source and use other procedures to manipulate data or regular Cypher to insert or update to the database.
There are also procedures for batching data, adding wait/sleep commands, and handling large data sets or temperamental data sources.

APOC's transformation procedures allow you to process dynamic labels or relationships, correct/skip null or empty values, format dates or other values, generate hashes, and handle other tricky data scenarios.

> While APOC is a flexible and custom data handling library, it may not be the most recommended method for more complex scenarios that could result in an excess of lines of code to handle multiple data transformations.

Reviewer: ```suggestion
While APOC is a flexible and custom data handling library, it is not always the recommended method for more complex scenarios that could result in an excess of lines of code to handle multiple data transformations.
```

== ETL Tool

> Neo4j's link:https://neo4j.com/labs/etl-tool/[ETL Tool] allows you to set up a JDBC connection to a relational database and then auto-map the data into a graph data model rendered as a visualization that you can edit according to your use case.

Reviewer: This is largely outdated, the ETL tool is no longer available to run in Desktop, for example. The blog post is from 2018, so it could be worth looking into whether this is actually something we should be calling out? It's still a Labs project.

Read the blog article link:https://medium.com/neo4j/tap-into-hidden-connections-translating-your-relational-data-to-graph-d3a2591d4026[Translating Relational Data to Graph^] for more information on how to do it with the ETL Tool.

== Apache Hop

With Apache Hop's link:https://hop.apache.org/manual/latest/technology/neo4j/index.html[Neo4j plugin^], you can load large datasets in Neo4j, update graphs, and get logging information.

== Drivers

> For importing data using a programming language, you can use the xref:languages-guides/index.adoc[Neo4j driver] for your preferred language and execute Cypher statements to/from the database.

Reviewer: ```suggestion
You can also import data programmatically using the xref:languages-guides/index.adoc[Neo4j driver] for your preferred language, executing Cypher statements directly against the database.
```

> This process is also helpful if you do not have access to the Cypher shell or if the data is not available as an accessible file.

Reviewer: ```suggestion
This approach is particularly useful when your data comes from an external API, a live application, or any source that isn't available as a flat file.
```

You can set up the driver connection to Neo4j, and then execute Cypher statements that pass from the application-level through the driver and to the database for various operations, such as including large amounts of inserts and updates.

Read the blog post https://medium.com/neo4j/5-tips-tricks-for-fast-batched-updates-of-graph-structures-with-neo4j-and-cypher-73c7f693c8cc[Tips and Tricks for Fast-Batched Import with Neo4j^] for more information on how to get started.

> You can retrieve data from a relational database (or other tabular structure) using xref:languages-guides/index.adoc[drivers] and the link:https://neo4j.com/docs/bolt/current/[Bolt Protocol].

Reviewer: ```suggestion
When you migrate data from a relational database (or other tabular structure) into Neo4j using the xref:languages-guides/index.adoc[Neo4j drivers], the driver communicates with the database over the link:https://neo4j.com/docs/bolt/current/[Bolt Protocol].
```
